### ISSUE TITLE
A potential bug in data normalization

### ISSUE THREAD

**Issue by patel-zeel** [awav/CGLB#3]
I am working on reproducing this paper and having the following concern:

[The paper](http://proceedings.mlr.press/v139/artemev21a/artemev21a.pdf) mentions that:
> Each input dimension is normalised to have mean 0 and variance 1 within the training set. Similarly, the training outputs are normalised to have 0 mean and variance 1. We apply the same normalisation to test data when making predictions, using the statistics computed on the training data.

I suppose that [the code corresponding to the above claim](https://github.com/awav/CGLB/blob/70060ca04eefbf7a810b06b8978bb4ccdfc44bfb/cglb_experiments/datasets.py#L68) is below (which is indeed correct):
```py
elif isinstance(data, utils.data.Dataset):
    train, test = (data.X_train, data.Y_train), (data.X_test, data.Y_test)
else:
    raise RuntimeError("Unknown dataset type detected")

(x_train, x_mu, x_std), (y_train, y_mu, y_std) = norm_dataset(train)
x_test = (test[0] - x_mu) / x_std
y_test = (test[1] - y_mu) / y_std
```

But, the entire data was already transformed in `bayesian_benchmarks` repo [over here](https://github.com/hughsalimbeni/bayesian_benchmarks/blob/a8fac6f285e700e29353208fb9828e7aa224aa4b/bayesian_benchmarks/data.py#L53).

```py
def normalize(X):
    X_mean = np.average(X, 0)[None, :]
    X_std = 1e-6 + np.std(X, 0)[None, :]
    return (X - X_mean) / X_std, X_mean, X_std

def preprocess_data(self, X, Y):
        X, self.X_mean, self.X_std = normalize(X)
        Y, self.Y_mean, self.Y_std = normalize(Y)
        return X, Y

class Dataset(object):
    def __init__(self, split=0, prop=0.9):
        if self.needs_download:
            self.download()

        X_raw, Y_raw = self.read_data()
        X, Y = self.preprocess_data(X_raw, Y_raw)
        ...
```

**Comment by awav** [comment]
This is not a problem, as we still end up with the same normalisation.

**Comment by patel-zeel** [comment]
Thanks for the quick response, @awav. But normalizing the entire train+test data beforehand is a bit unrealistic setting right? because in real-life we do not have access to the test data beforehand.

**Comment by awav** [comment]
> But normalizing the entire train+test data beforehand is a bit unrealistic setting right? 

We normalize datasets after loading them using bayesian benchmarks, it is consistent with the paper description.
